### DISCUSSION TITLE
GitHub Pages serving .mp4 files with incorrect MIME type

### DISCUSSION THREAD

**Discussion by gosko** [community/community#58259] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Bug

### Body

GitHub Pages is serving `.mp4` files with MIME type `application/mp4`, for example [this file](https://w3c.github.io/wai-website/content-assets/wcag-act-rules/test-assets/rabbit-video/video.mp4).

They should be served as `video/mp4`.

References:
- [MDN: MIME types (IANA media types)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types)
- [MDN: Common Mime Types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types)
- [MP4 file format in wikipedia](https://en.wikipedia.org/wiki/MP4_file_format)
- [Apache server config](https://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types)
- [nginx server config](https://github.com/nginx/nginx/blob/master/conf/mime.types)

According to the [GitHub Pages documentation](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#mime-types-on-github-pages) these types are coming from the [mime-db project](https://github.com/jshttp/mime-db) but [the db.json file from that project](https://github.com/jshttp/mime-db/blob/master/db.json) has both `application/mp4` and `video/mp4` listed for `.mp4` file extensions. I'm not sure why GitHub Pages has chosen to interpret that to mean `.mp4` files should be served as `application/mp4` instead of `video/mp4`.

**Comment by the-officialjosh** [comment]
To the best of my Knowledge - ~~According to the GitHub Pages documentation, .mp4 files are served with the MIME type application/mp4 instead of video/mp4. While the mime-db project lists both MIME types for .mp4 files, GitHub Pages have chosen to interpret it as application/mp4 instead of video/mp4. This could be due to their specific configuration and handling of MIME types.~~
If you believe that .mp4 files should be served as video/mp4 on GitHub Pages, you can consider reaching out to GitHub support or opening an issue on the [GitHub Pages repository](https://github.com/github/pages-gem) to inquire about their reasoning behind the choice of MIME type and to request a potential change. Additionally, you can refer to the provided references for more information on MIME types and server configurations.

**Reply by gosko** [reply]
Where does the GitHub Pages documentation say that .mp4 files are served with MIME type application/mp4 instead of video/mp4?

**Reply by the-officialjosh** [reply]
I saw that from another issue on GitHub ([source](https://github.com/isaacs/github/issues/1628)) - 

A proposed solution is to add support for a default MIME type in GitHub Pages. One possible approach is to include the following code in the `_config.yml` file:

```yaml
github:
  default_mime_type: video/mp4
```

However, it remains uncertain whether this solution has been implemented or not.

**Reply by gosko** [reply]
No, that's not a workable solution

**Comment by yoannchaudet** [comment]
GitHub Pages uses https://github.com/jshttp/mime-db for MIME type. If you need `video/mp4` instead of `application/mp4` you will need to open an issue and justify it in their repo!

**Reply by gosko** [reply]
As I mentioned, mime-db has both application/mp4 and video/mp4 listed for .mp4 file extensions. The issue is that GitHub Pages has chosen to use application/mp4 instead of video/mp4 for some reason.

**Reply by gosko** [reply]
mime-db has [this for application/mp4](https://github.com/jshttp/mime-db/blob/a76e5a824c228e2e58363c9404e42a54ee1d142f/db.json#L1046-L1049):
```
  "application/mp4": {
    "source": "iana",
    "extensions": ["mp4","mpg4","mp4s","m4p"]
  },
```

and [this for video/mp4](https://github.com/jshttp/mime-db/blob/a76e5a824c228e2e58363c9404e42a54ee1d142f/db.json#L8522-L8526):

```
  "video/mp4": {
    "source": "iana",
    "compressible": false,
    "extensions": ["mp4","mp4v","mpg4"]
  },
```

so GitHub Pages must have some logic to select one or the other when there are multiple types specified for a given file extension.

**Reply by yoannchaudet** [reply]
Oh my mistake I did not catch that you were aware we used mime-db!

Looked briefly. The `mp4` extension is ambiguous and you have to open and inspect the file to decide if it should be served as `application/mp4`, `video/mp4` or `audio/mp4` (as per the [spec](https://www.rfc-editor.org/rfc/rfc4337.txt)). We obviously don't do that today (no plans) and we don't support custom headers (yet).

If you want explicitly `video/mp4`, you could just rename your file to use the `mp4v` extension which we support and which is not ambiguous.

**Reply by gosko** [reply]
Hi @yoannchaudet, thanks for looking into this!

I understand `.mp4` can be ambiguous but for the vast majority of cases in common use `video/mp4` is the best option which is why it's the default almost everywhere else (references: [apache](https://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types), [nginx](https://github.com/nginx/nginx/blob/master/conf/mime.types), [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types), [wikipedia](https://en.wikipedia.org/wiki/MP4_file_format#Filename_extensions), [debian](https://salsa.debian.org/debian/media-types/-/blob/master/mime.types#L2241), [ubuntu](https://git.launchpad.net/ubuntu/+source/media-types/tree/mime.types#n2217), [freedesktop.org](https://gitlab.freedesktop.org/xdg/shared-mime-info/-/blob/master/data/freedesktop.org.xml.in#L4485))

I will see if our users can use `.mp4v` as a workaround but I expect this will continue to be an issue for users of GitHub Pages since the rest of the web has settled on `.mp4` == `video/mp4`.

**Comment by gosko** [comment, marked as answer]
This appears to have been fixed: GitHub Pages is now returning `.mp4` files as `video/mp4`. [Related discussion elsewhere](https://github.com/khellang/MimeTypes/issues/18#issuecomment-1601728654).
